src/annotator/anchoring/range.js

```diff
@@ -0,0 +1,497 @@
+  /**
+   * Public: normalize works around the fact that browsers don't generate
+   * ranges/selections in a consistent manner. Some (Safari) will create
+   * ranges that have (say) a textNode startContainer and elementNode
+   * endContainer. Others (Firefox) seem to only ever generate
+   * textNode/textNode or elementNode/elementNode pairs.
+   *
+   * Returns an instance of NormalizedRange
+   */
+  normalize() {
+    if (this.tainted) {
+      throw new Error('You may only call normalize() once on a BrowserRange!');
+    } else {
+      this.tainted = true;
+    }
+    const range = {};
+
+    // Look at the start
+    if (this.startContainer.nodeType === Node.ELEMENT_NODE) {
+      // We are dealing with element nodes
+      if (this.startOffset < this.startContainer.childNodes.length) {
+        range.start = getFirstTextNodeNotBefore(
+          this.startContainer.childNodes[this.startOffset]
+        );
+      } else {
+        range.start = getFirstTextNodeNotBefore(this.startContainer);
+      }
+      range.startOffset = 0;
+    } else {
+      // We are dealing with simple text nodes
+      range.start = this.startContainer;
+      range.startOffset = this.startOffset;
+    }
+
+    // Look at the end
+    if (this.endContainer.nodeType === Node.ELEMENT_NODE) {
+      // Get specified node.
+      let node = this.endContainer.childNodes[this.endOffset];
+      // Does that node exist?
+      if (node) {
+        // Look for a text node either at the immediate beginning of node
+        let n = node;
+        while (n && n.nodeType !== Node.TEXT_NODE) {
+          n = n.firstChild;
+        }
+        // Did we find a text node at the start of this element?
+        if (n) {
+          range.end = n;
+          range.endOffset = 0;
+        }
+      }
+
+      if (!range.end) {
+        // We need to find a text node in the previous sibling of the node at the
+        // given offset, if one exists, or in the previous sibling of its container.
+        if (this.endOffset) {
+          node = this.endContainer.childNodes[this.endOffset - 1];
+        } else {
+          node = this.endContainer.previousSibling;
+        }
+        range.end = getLastTextNodeUpTo(node);
+        range.endOffset = range.end.nodeValue.length;
+      }
+    } else {
+      // We are dealing with simple text nodes
+      range.end = this.endContainer;
+      range.endOffset = this.endOffset;
+    }
+
+    // We have collected the initial data.
+    // Now let's start to slice & dice the text elements!
+    const normalRange = {};
+
+    if (range.startOffset > 0) {
+      // Do we really have to cut?
+      if (
+        !range.start.nextSibling ||
+        range.start.nodeValue.length > range.startOffset
+      ) {
+        // Yes. Cut.
+        normalRange.start = range.start.splitText(range.startOffset);
+      } else {
+        // Avoid splitting off zero-length pieces.
+        normalRange.start = getFirstTextNodeNotBefore(range.start.nextSibling);
```

Comment:
This was not a straight conversion, the original coffeescript did not call getFirstTextNodeNotBefore(), it just set normalRange.start to rangeTemp.start.nextSibling. but nextSibling may not always be a text node. I discovered this issue while building out unit tests. I'm not 100% I'm right about this, but it could be a long standing bug, or just my misunderstanding.

Reply:
What was the test case?

Reply:
Yes, to the best of my knowledge when flushing out the unit tests for complete coverage, I came across a case were the  range.start.nextSibling was not sufficient. I think it was pointing to a <br> tag but it really needed to look before that to the next node.


The test on line 205: "does not cut the text node" will execute this code. If you remove the function call `getFirstTextNodeNotBefore` then the test will fail, but its the way that it failed that I was suspicious of because `normalizedRange.start` (in that case)  points to the <br> tag. I suspect it needed to actually point to a text node which is why I made the change and added the `getFirstTextNodeNotBefore` call.

Reply:
I see. So it looks like this happens when the range to be normalized starts at the end of a text node? OK, this would be a legitimate range that a browser might create as part of the selection.